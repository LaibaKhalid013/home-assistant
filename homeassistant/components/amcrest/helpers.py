"""Helpers for amcrest component."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.helpers.typing import UndefinedType


def service_signal(service: str, *args: str) -> str:
    """Encode signal."""
    return "_".join([DOMAIN, service, *args])


def log_update_error(
    logger: logging.Logger,
    action: str,
    name: str | UndefinedType | None,
    entity_type: str,
    error: Exception,
    level: int = logging.ERROR,
) -> None:
    """Log an update error."""
    logger.log(
        level,
        "Could not %s %s %s due to error: %s",
        action,
        name,
        entity_type,
        error.__class__.__name__,
    )
