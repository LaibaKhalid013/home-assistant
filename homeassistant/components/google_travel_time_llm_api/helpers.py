"""Helpers for Google Time Travel integration."""

import logging

from googlemaps import Client

from homeassistant.core import HomeAssistant

_LOGGER = logging.getLogger(__name__)


def validate_config_entry(hass: HomeAssistant, api_key: str) -> None:
    """Return whether the config entry data is valid."""
    try:
        Client(api_key, timeout=10)
    except ValueError as value_error:
        _LOGGER.error("Malformed API key")
        raise InvalidApiKeyException from value_error


class InvalidApiKeyException(Exception):
    """Invalid API Key Error."""
