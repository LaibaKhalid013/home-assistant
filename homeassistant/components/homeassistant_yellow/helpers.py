"""Helper functions for the Yellow."""

from __future__ import annotations

import asyncio
import enum

import usb

from homeassistant.core import HomeAssistant


class YellowGPIO(enum.IntEnum):
    """Yellow MGM210P GPIO pins."""

    RADIO_SWDIO = 6
    RADIO_SWCLK = 7
    RADIO_TXD = 8
    RADIO_RXD = 9
    RADIO_CTS = 10
    RADIO_RTS = 11
    RADIO_BOOT = 24
    RADIO_RESET = 25
    SW_USER = 26
    SW_WIPE = 27


# Pin states on a properly installed CM4
RUNNING_PIN_STATES = {
    YellowGPIO.RADIO_BOOT: True,
    YellowGPIO.RADIO_RESET: True,
}

GPIO_READ_ATTEMPTS = 5
GPIO_READ_DELAY_S = 0.01


# Bus 001 Device 001: ID 1d6b:0002
CM4_USB_HUB_VENDOR = 0x1D6B
CM4_USB_HUB_PRODUCT = 0x0002

# Bus 001 Device 002: ID 1a40:0101
YELLOW_USB_HUB_VENDOR = 0x1A40
YELLOW_USB_HUB_PRODUCT = 0x0101


def _read_gpio_pins[_PinsT: int](pins: list[_PinsT]) -> dict[_PinsT, bool]:
    """Read the state of the given GPIO pins."""
    # pylint: disable-next=import-outside-toplevel
    import gpiod

    chip = gpiod.chip("/dev/gpiochip0", gpiod.chip.OPEN_BY_PATH)
    lines = chip.get_lines(pins)

    config = gpiod.line_request()
    config.consumer = "core-yellow"
    config.request_type = gpiod.line_request.DIRECTION_INPUT

    try:
        lines.request(config)
        return {p: bool(v) for p, v in zip(pins, lines.get_values(), strict=True)}
    finally:
        lines.release()


async def async_validate_gpio_states(hass: HomeAssistant) -> bool:
    """Validate the state of the GPIO pins."""
    # We read multiple times in a row to make sure the GPIO state isn't fluctuating
    for attempt in range(GPIO_READ_ATTEMPTS):
        if attempt > 0:
            await asyncio.sleep(GPIO_READ_DELAY_S)

        pin_states = await hass.async_add_executor_job(
            _read_gpio_pins, list(RUNNING_PIN_STATES)
        )

        if pin_states != RUNNING_PIN_STATES:
            return False

    return True


def validate_usb_hub_present() -> bool:
    """Validate that the USB hub is present."""
    root_hub = usb.core.find(
        # The root hub has no parent
        parent=None,
        idVendor=CM4_USB_HUB_VENDOR,
        idProduct=CM4_USB_HUB_PRODUCT,
    )

    # This really should not be possible
    if root_hub is None:
        return False

    yellow_hub = usb.core.find(
        # The Yellow's two-port hub is a child of the pi's root hub
        parent=root_hub,
        idVendor=YELLOW_USB_HUB_VENDOR,
        idProduct=YELLOW_USB_HUB_PRODUCT,
    )

    return yellow_hub is not None


async def async_validate_hardware_consistent(hass: HomeAssistant) -> bool:
    """Validate the hardware is consistent with a properly installed CM4."""
    if not await hass.async_add_executor_job(validate_usb_hub_present):
        return False

    if not await async_validate_gpio_states(hass):
        return False

    return True
