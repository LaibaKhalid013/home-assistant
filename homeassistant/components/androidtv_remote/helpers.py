"""Helper functions for Android TV Remote integration."""

from __future__ import annotations

from typing import TYPE_CHECKING

from androidtvremote2 import AndroidTVRemote

from homeassistant.helpers.storage import STORAGE_DIR

from .const import CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant


def create_api(hass: HomeAssistant, host: str, enable_ime: bool) -> AndroidTVRemote:
    """Create an AndroidTVRemote instance."""
    return AndroidTVRemote(
        client_name="Home Assistant",
        certfile=hass.config.path(STORAGE_DIR, "androidtv_remote_cert.pem"),
        keyfile=hass.config.path(STORAGE_DIR, "androidtv_remote_key.pem"),
        host=host,
        loop=hass.loop,
        enable_ime=enable_ime,
    )


def get_enable_ime(entry: ConfigEntry) -> bool:
    """Get value of enable_ime option or its default value."""
    return entry.options.get(CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE)
