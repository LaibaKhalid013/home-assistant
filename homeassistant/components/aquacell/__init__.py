"""The Aquacell integration."""
from __future__ import annotations

from aioaquacell import AquacellApi

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DATA_AQUACELL, DOMAIN
from .coordinator import AquacellCoordinator

PLATFORMS = [Platform.SENSOR]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Aquacell from a config entry."""
    session = async_get_clientsession(hass)

    aquacell_api = AquacellApi(session)

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {DATA_AQUACELL: aquacell_api}

    coordinator = AquacellCoordinator(hass, aquacell_api, entry)

    await coordinator.async_config_entry_first_refresh()
    hass.data[DOMAIN][entry.entry_id] = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
