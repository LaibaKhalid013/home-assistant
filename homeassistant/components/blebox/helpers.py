"""Blebox helpers."""

from __future__ import annotations

from typing import TYPE_CHECKING

import aiohttp

from homeassistant.helpers.aiohttp_client import (
    async_create_clientsession,
    async_get_clientsession,
)

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant


def get_maybe_authenticated_session(
    hass: HomeAssistant, password: str | None, username: str | None
) -> aiohttp.ClientSession:
    """Return proper session object."""
    if username and password:
        auth = aiohttp.BasicAuth(login=username, password=password)
        return async_create_clientsession(hass, auth=auth)

    return async_get_clientsession(hass)
