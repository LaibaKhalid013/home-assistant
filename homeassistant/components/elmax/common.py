"""Elmax integration common classes and utilities."""

from __future__ import annotations

import ssl

from elmax_api.model.endpoint import DeviceEndpoint
from elmax_api.model.panel import PanelEntry
from packaging import version

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import (
    DOMAIN,
    ELMAX_LOCAL_API_PATH,
    MIN_APIV2_SUPPORTED_VERSION,
    SIGNAL_PANEL_UPDATE,
)
from .coordinator import ElmaxCoordinator


def get_direct_api_url(host: str, port: int, use_ssl: bool) -> str:
    """Return the direct API url given the base URI."""
    schema = "https" if use_ssl else "http"
    return f"{schema}://{host}:{port}/{ELMAX_LOCAL_API_PATH}"


def build_direct_ssl_context(cadata: str) -> ssl.SSLContext:
    """Create a custom SSL context for direct-api verification."""
    context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
    context.check_hostname = False
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_verify_locations(cadata=cadata)
    return context


def check_local_version_supported(api_version: str | None) -> bool:
    """Check whether the given API version is supported."""
    if api_version is None:
        return False
    return version.parse(api_version) >= version.parse(MIN_APIV2_SUPPORTED_VERSION)


class DirectPanel(PanelEntry):
    """Helper class for wrapping a directly accessed Elmax Panel."""

    def __init__(self, panel_uri):
        """Construct the object."""
        super().__init__(panel_uri, True, {})

    def get_name_by_user(self, username: str) -> str:
        """Return the panel name."""
        return f"Direct Panel {self.hash}"


class ElmaxEntity(CoordinatorEntity[ElmaxCoordinator]):
    """Wrapper for Elmax entities."""

    _last_state: DeviceEndpoint = None

    def __init__(
        self,
        elmax_device: DeviceEndpoint,
        panel_version: str,
        coordinator: ElmaxCoordinator,
    ) -> None:
        """Construct the object."""
        super().__init__(coordinator=coordinator)
        self._device = elmax_device
        self._attr_unique_id = elmax_device.endpoint_id
        self._attr_name = elmax_device.name
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, coordinator.panel_entry.hash)},
            name=coordinator.panel_entry.get_name_by_user(
                coordinator.http_client.get_authenticated_username()
            ),
            manufacturer="Elmax",
            model=panel_version,
            sw_version=panel_version,
        )

    async def async_added_to_hass(self) -> None:
        """Register push notifications callbacks if available."""
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                f"{SIGNAL_PANEL_UPDATE}-{self.coordinator.panel_entry.hash}-{self._device.endpoint_id}",
                self._handle_update,
            )
        )
        return await super().async_added_to_hass()

    @callback
    def _handle_update(self, endpoint_status: DeviceEndpoint) -> None:
        self._last_state = endpoint_status
        self.async_write_ha_state()

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self.coordinator.panel_entry.online
