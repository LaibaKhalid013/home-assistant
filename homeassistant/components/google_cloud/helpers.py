"""Helper classes for Google Cloud integration."""

from __future__ import annotations

import logging
from types import MappingProxyType
from typing import Any

from google.cloud import texttospeech
import voluptuous as vol

from homeassistant.components.tts import CONF_LANG
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.selector import (
    NumberSelector,
    NumberSelectorConfig,
    SelectSelector,
    SelectSelectorConfig,
    SelectSelectorMode,
)

from .const import (
    CONF_ENCODING,
    CONF_GAIN,
    CONF_GENDER,
    CONF_KEY_FILE,
    CONF_PITCH,
    CONF_PROFILES,
    CONF_SPEED,
    CONF_TEXT_TYPE,
    CONF_VOICE,
    DEFAULT_LANG,
)

_LOGGER = logging.getLogger(__name__)

DEFAULT_GENDER = texttospeech.SsmlVoiceGender.NEUTRAL.name
DEFAULT_VOICE = ""
DEFAULT_ENCODING = texttospeech.AudioEncoding.MP3.name
DEFAULT_SPEED = 1.0
DEFAULT_PITCH = 0
DEFAULT_GAIN = 0
DEFAULT_PROFILES: list[str] = []
DEFAULT_TEXT_TYPE = "text"

# https://cloud.google.com/text-to-speech/docs/audio-profiles
SUPPORTED_PROFILES = [
    "wearable-class-device",
    "handset-class-device",
    "headphone-class-device",
    "small-bluetooth-speaker-class-device",
    "medium-bluetooth-speaker-class-device",
    "large-home-entertainment-class-device",
    "large-automotive-class-device",
    "telephony-class-application",
]
SUPPORTED_TEXT_TYPES = ["text", "ssml"]


async def async_tts_voices(
    client: texttospeech.TextToSpeechAsyncClient,
) -> dict[str, list[str]]:
    """Get TTS voice models keyed by language."""
    voices: dict[str, list[str]] = {}
    list_voices_response = await client.list_voices()
    for voice in list_voices_response.voices:
        language_code = voice.language_codes[0]
        if language_code not in voices:
            voices[language_code] = []
        voices[language_code].append(voice.name)
    return voices


def tts_options_schema(
    config_options: MappingProxyType[str, Any], voices: dict[str, list[str]]
):
    """Return schema for TTS options with default values from config or constants."""
    return vol.Schema(
        {
            vol.Optional(
                CONF_GENDER,
                description={"suggested_value": config_options.get(CONF_GENDER)},
                default=DEFAULT_GENDER,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=list(texttospeech.SsmlVoiceGender.__members__),
                )
            ),
            vol.Optional(
                CONF_VOICE,
                description={"suggested_value": config_options.get(CONF_VOICE)},
                default=DEFAULT_VOICE,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=["", *sum(voices.values(), [])],
                )
            ),
            vol.Optional(
                CONF_ENCODING,
                description={"suggested_value": config_options.get(CONF_ENCODING)},
                default=DEFAULT_ENCODING,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=list(texttospeech.AudioEncoding.__members__),
                )
            ),
            vol.Optional(
                CONF_SPEED,
                description={"suggested_value": config_options.get(CONF_SPEED)},
                default=DEFAULT_SPEED,
            ): NumberSelector(NumberSelectorConfig(min=0.25, max=4, step=0.05)),
            vol.Optional(
                CONF_PITCH,
                description={"suggested_value": config_options.get(CONF_PITCH)},
                default=DEFAULT_PITCH,
            ): NumberSelector(NumberSelectorConfig(min=-20, max=20, step=0.1)),
            vol.Optional(
                CONF_GAIN,
                description={"suggested_value": config_options.get(CONF_GAIN)},
                default=DEFAULT_GAIN,
            ): NumberSelector(NumberSelectorConfig(min=-96, max=16, step=1)),
            vol.Optional(
                CONF_PROFILES,
                description={"suggested_value": config_options.get(CONF_PROFILES)},
                default=DEFAULT_PROFILES,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=SUPPORTED_PROFILES,
                    multiple=True,
                    sort=False,
                )
            ),
            vol.Optional(
                CONF_TEXT_TYPE,
                description={"suggested_value": config_options.get(CONF_TEXT_TYPE)},
                default=DEFAULT_TEXT_TYPE,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=SUPPORTED_TEXT_TYPES,
                )
            ),
        }
    )


def tts_platform_schema():
    """Return schema for TTS platform."""
    return vol.Schema(
        {
            vol.Optional(CONF_KEY_FILE): cv.string,
            vol.Optional(CONF_LANG, default=DEFAULT_LANG): cv.matches_regex(
                r"[a-z]{2,3}-[A-Z]{2}"
            ),
            **tts_options_schema(config_options={}, voices={}).schema,
            vol.Optional(CONF_VOICE, default=DEFAULT_VOICE): cv.matches_regex(
                r"[a-z]{2,3}-[A-Z]{2}-.*-[A-Z]|"
            ),
        }
    )
