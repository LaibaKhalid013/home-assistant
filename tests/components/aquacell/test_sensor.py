"""Test the Aquacell init module."""

from __future__ import annotations

from unittest.mock import AsyncMock

import pytest

from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from tests.common import MockConfigEntry
from tests.components.aquacell import DSN, setup_integration


@pytest.mark.parametrize(
    ("sensor_id"),
    [
        ("wi_fi_level"),
    ],
)
async def test_sensors(
    hass: HomeAssistant,
    mock_aquacell_api: AsyncMock,
    mock_config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
    sensor_id: str,
) -> None:
    """Test the creation of Aquacell sensors."""
    await setup_integration(hass, mock_config_entry)
    entity_id = f"sensor.aquacell_name_{sensor_id}"

    assert entity_registry.async_is_registered(entity_id)
    entry = entity_registry.async_get(entity_id)
    assert entry.unique_id == f"{DSN}-{sensor_id}"
    assert hass.states.get(entity_id).state == "high"
