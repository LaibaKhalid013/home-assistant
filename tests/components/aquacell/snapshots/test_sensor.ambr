# serializer version: 1
# name: test_sensors[sensor.aquacell_dsn_battery-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.aquacell_dsn_battery',
    'has_entity_name': False,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.BATTERY: 'battery'>,
    'original_icon': None,
    'original_name': None,
    'platform': 'aquacell',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'DSN-battery',
    'unit_of_measurement': '%',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_battery-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'battery',
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.aquacell_dsn_battery',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '40',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_left_side_percentage-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.aquacell_dsn_salt_left_side_percentage',
    'has_entity_name': False,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'aquacell',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'salt_left_side_percentage',
    'unique_id': 'DSN-salt_left_side_percentage',
    'unit_of_measurement': '%',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_left_side_percentage-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.aquacell_dsn_salt_left_side_percentage',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '100',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_left_side_time_remaining-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.aquacell_dsn_salt_left_side_time_remaining',
    'has_entity_name': False,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.DURATION: 'duration'>,
    'original_icon': None,
    'original_name': None,
    'platform': 'aquacell',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'salt_left_side_time_remaining',
    'unique_id': 'DSN-salt_left_side_time_remaining',
    'unit_of_measurement': <UnitOfTime.DAYS: 'd'>,
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_left_side_time_remaining-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'duration',
      'unit_of_measurement': <UnitOfTime.DAYS: 'd'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.aquacell_dsn_salt_left_side_time_remaining',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '30',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_right_side_percentage-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.aquacell_dsn_salt_right_side_percentage',
    'has_entity_name': False,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'aquacell',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'salt_right_side_percentage',
    'unique_id': 'DSN-salt_right_side_percentage',
    'unit_of_measurement': '%',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_right_side_percentage-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.aquacell_dsn_salt_right_side_percentage',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '100',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_right_side_time_remaining-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.aquacell_dsn_salt_right_side_time_remaining',
    'has_entity_name': False,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.DURATION: 'duration'>,
    'original_icon': None,
    'original_name': None,
    'platform': 'aquacell',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'salt_right_side_time_remaining',
    'unique_id': 'DSN-salt_right_side_time_remaining',
    'unit_of_measurement': <UnitOfTime.DAYS: 'd'>,
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_salt_right_side_time_remaining-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'duration',
      'unit_of_measurement': <UnitOfTime.DAYS: 'd'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.aquacell_dsn_salt_right_side_time_remaining',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '30',
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_wi_fi_strength-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        'high',
        'medium',
        'low',
      ]),
    }),
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.aquacell_dsn_wi_fi_strength',
    'has_entity_name': False,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.ENUM: 'enum'>,
    'original_icon': None,
    'original_name': None,
    'platform': 'aquacell',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'wi_fi_strength',
    'unique_id': 'DSN-wi_fi_strength',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensors[sensor.aquacell_dsn_wi_fi_strength-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'enum',
      'options': list([
        'high',
        'medium',
        'low',
      ]),
    }),
    'context': <ANY>,
    'entity_id': 'sensor.aquacell_dsn_wi_fi_strength',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'high',
  })
# ---
